### DISCUSSION TITLE
Seeing hang in SSL_read() even after disabling SSL_MODE_AUTO_RETRY

### DISCUSSION THREAD

**Discussion by srahul292919** [openssl/openssl#22546] [Q&A] [answered]
Hi All,
Recently, we had added support for TLSv1.3 in our product.
To support TLSv1.3 we did the following changes.
1. We had to switch off the SSL_MODE_AUTO_RETRY as we were seeing a hang with blocking sockets when SSL_read() was processing the non-application data records, but then try to read an application data record which might not be available.
2. We added the SSL_read() call with 0 bytes, just after the handshake is completed, to read the session ticket if it is available.
3. We had also changed the read logic to handle SSL_ERROR_WANT_READ, SSL_ERROR_WANT_WRITE errors during the read.
 
Now, we are observing the following hang for the 2nd step where the SSL_read() was called with 0 bytes, just after the handshake is completed, to read the session ticket if it is available.
 
#0  0x00007fd34a5886e0 in __read_nocancel () at ../sysdeps/unix/syscall-template.S:81
#1  0x0000561992233735 in sock_read ()
#2  0x000056199223052c in bread_conv ()
#3  0x000056199222f222 in BIO_read ()
#4  0x00005619921cd190 in ssl3_read_n ()
#5  0x00005619921d1996 in ssl3_get_record ()
#6  0x00005619921cf233 in ssl3_read_bytes ()
#7  0x00005619921d5e4a in ssl3_read ()
#8  0x00005619921defed in ssl_read_internal ()
#9  0x00005619921df1d3 in SSL_read ()
 ....
 
Any idea why this hang is coming even after disabling SSL_MODE_AUTO_RETRY.
Let me know if you need any other details. Thanks in advance.

**Comment by mattcaswell** [comment, marked as answer]
If you call `SSL_read()` then it will try to read a record from the underlying BIO. If the BIO is using a blocking socket and there is no data to read yet then it will block. It will do this whether or not you are attempting to read "0" bytes.

> Now, we are observing the following hang for the 2nd step where the SSL_read() was called with 0 bytes, just after the handshake is completed, to read the session ticket if it is available.

So, in this scenario, you don't know whether or not there is a session ticket, or any other kind of data on the underlying socket. Calling `SSL_read` with a length of 0 will still try to read a record. If there is no session ticket, or anything else, then it will hang if you are using a blocking socket.

To avoid this you should check whether the underlying socket is readable or not. Don't call `SSL_read` if the underlying socket is not readable. Even this doesn't guarantee that it won't block. It is theoretically possible for "half" a record to be available, in which case it will block until the whole record arrives.

**Reply by srahul292919** [reply]
_**Really appreciate the quick reply @mattcaswell  !!**_

If you call SSL_read() then it will try to read a record from the underlying BIO. If the BIO is using a blocking socket and there is no data to read yet then it will block. It will do this whether or not you are attempting to read "0" bytes.
**_<Is this valid with after removing SSL_MODE_AUTO_RETRY mode. If yes, what is the exact purpose SSL_MODE_AUTO_RETRY flag.>_**

So, in this scenario, you don't know whether or not there is a session ticket, or any other kind of data on the underlying socket. Calling SSL_read with a length of 0 will still try to read a record. If there is no session ticket, or anything else, then it will hang if you are using a blocking socket.
__<I was under impression that if SSL_read() is called with length 0 and SSL_MODE_AUTO_RETRY mode is removed, it should try to read the 0 length application data and read non application data (session ticket) if available.  >__


To avoid this you should check whether the underlying socket is readable or not. Don't call SSL_read if the underlying socket is not readable. Even this doesn't guarantee that it won't block. It is theoretically possible for "half" a record to be available, in which case it will block until the whole record arrives.
**_< I added a select() call to see if there is any activity on the socket before calling SSL_read(). Looks like there is some activity on that socket and SSL_read() (which used to hang earlier) is returning SSL_ERROR_WANT_READ error now. When I call SSL_read() again to read the left over data, it is returning SSL_ERROR_SYSCALL error. Not sure what is happening here. ERR_get_error() is returning 0 and ERR_error_string() returns 1453792608. Do you have any idea about this error code. Or do you have any sample code for this which I can refer to. >_**

**Reply by srahul292919** [reply]
Hi @mattcaswell , can you please help me with the follow up questions. Let me know if you need any other details.

**Reply by mattcaswell** [reply]
> If you call SSL_read() then it will try to read a record from the underlying BIO. If the BIO is using a blocking socket and there is no data to read yet then it will block. It will do this whether or not you are attempting to read "0" bytes.
<Is this valid with after removing SSL_MODE_AUTO_RETRY mode. If yes, what is the exact purpose SSL_MODE_AUTO_RETRY flag.>

Yes, this is true regardless of the `SSL_MODE_AUTO_RETRY` mode setting. The `SSL_MODE_AUTO_RETRY` mode changes what OpenSSL does *after* it has read a record that did not yield any application data, i.e. if you call `SSL_read()` and `SSL_MODE_AUTO_RETRY` is enabled, then if a non app-data record is read then OpenSSL will immediately try and read another record (and continue to do that until it has read a record that does contain app data).

> So, in this scenario, you don't know whether or not there is a session ticket, or any other kind of data on the underlying socket. Calling SSL_read with a length of 0 will still try to read a record. If there is no session ticket, or anything else, then it will hang if you are using a blocking socket.
<I was under impression that if SSL_read() is called with length 0 and SSL_MODE_AUTO_RETRY mode is removed, it should try to read the 0 length application data and read non application data (session ticket) if available. >

If you try and read with length 0 using a blocking socket, and there is no SSL_MODE_AUTO_RETRY set, then it will attempt to read a record from the underlying BIO. If no such record is available yet then it will block until a record is available. Once a record is read then, if it is non-app data, the non-app data record will be processed and SSL_read() will return. If the record contains app-data then it will be buffered and SSL_read() will still return. A subsequent call to SSL_read() with a non zero length will return the buffered data.

 > < I added a select() call to see if there is any activity on the socket before calling SSL_read(). Looks like there is some activity on that socket and SSL_read() (which used to hang earlier) is returning SSL_ERROR_WANT_READ error now. When I call SSL_read() again to read the left over data, it is returning SSL_ERROR_SYSCALL error. Not sure what is happening here. ERR_get_error() is returning 0 and ERR_error_string() returns 1453792608. Do you have any idea about this error code. Or do you have any sample code for this which I can refer to. >

I don't know why you are seeing `SSL_ERROR_SYSCALL`. If you are getting that then you should consult errno for further information.
